File: nationseal/jsondb.py
"""In-memory JSON file database. Keeps the whole dataset in memory and flushes on mutation."""

from __future__ import annotations

import json
import uuid
from datetime import UTC
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
	from nationseal.models import (
		DatabaseSchema,
		GuildState,
		Reviewer,
		Sanction,
		SanctionRequest,
		SanctionRequestStatus,
	)

DEFAULT_SCHEMA: DatabaseSchema = {
	"sanctionRequests": [],
	"sanctions": [],
	"reviewers": [],
	"guildStates": [],
}


class JsonDatabase:
	def __init__(self, file_path: str) -> None:
		self.file_path = Path(file_path)
		self.data: DatabaseSchema = {**DEFAULT_SCHEMA}
		self.loaded = False
		self.modified = False
		self._flush_promise: object | None = None

	async def connect(self) -> None:
		if self.loaded:
			return
		if self.file_path.exists():
			raw = self.file_path.read_text(encoding="utf-8")
			parsed = json.loads(raw)
			self.data = {**DEFAULT_SCHEMA, **parsed}
			self.modified = False
		else:
			self.data = {**DEFAULT_SCHEMA}
			self.modified = False
			self.flush()
		self.loaded = True

	def flush(self) -> None:
		if not self.modified:
			return
		self.file_path.parent.mkdir(parents=True, exist_ok=True)
		self.file_path.write_text(json.dumps(self.data, indent="\t"), encoding="utf-8")
		self.modified = False

	def touch(self) -> None:
		self.modified = True

	# Sanction requests

	async def create_sanction_request(self, data: SanctionRequest) -> SanctionRequest:
		record: SanctionRequest = {**data, "id": str(uuid.uuid4())}
		self.data["sanctionRequests"].append(record)
		self.touch()
		self.flush()
		return record

	async def get_sanction_request(self, request_id: str) -> SanctionRequest | None:
		for request in self.data["sanctionRequests"]:
			if request["id"] == request_id:
				return request
		return None

	async def update_sanction_request(
		self, request_id: str, patch: dict[str, object]
	) -> SanctionRequest:
		for index, request in enumerate(self.data["sanctionRequests"]):
			if request["id"] == request_id:
				updated: SanctionRequest = {**request, **patch, "id": request_id}  # type: ignore[typeddict-item]
				self.data["sanctionRequests"][index] = updated
				self.touch()
				self.flush()
				return updated
		raise ValueError(f"Sanction request {request_id} not found")

	async def list_sanction_requests_by_status(
		self, status: SanctionRequestStatus, limit: int
	) -> list[SanctionRequest]:
		filtered = [r for r in self.data["sanctionRequests"] if r["status"] == status]
		filtered.sort(key=lambda r: r["submittedAt"], reverse=True)
		return filtered[:limit]

	# Sanctions

	async def get_sanction(self, target_id: str) -> Sanction | None:
		for sanction in self.data["sanctions"]:
			if sanction["id"] == target_id:
				return sanction
		return None

	async def upsert_sanction(self, target_id: str, data: Sanction) -> Sanction:
		record: Sanction = {**data, "id": target_id}
		for index, sanction in enumerate(self.data["sanctions"]):
			if sanction["id"] == target_id:
				self.data["sanctions"][index] = record
				self.touch()
				self.flush()
				return record
		self.data["sanctions"].append(record)
		self.touch()
		self.flush()
		return record

	async def remove_sanction(self, target_id: str) -> None:
		self.data["sanctions"] = [s for s in self.data["sanctions"] if s["id"] != target_id]
		self.touch()
		self.flush()

	async def list_all_sanctions(self) -> list[Sanction]:
		return list(self.data["sanctions"])

	# Reviewers

	async def is_reviewer(self, user_id: str) -> bool:
		return any(r["id"] == user_id for r in self.data["reviewers"])

	async def add_reviewer(self, user_id: str, added_by: str) -> Reviewer:
		existing = next((r for r in self.data["reviewers"] if r["id"] == user_id), None)
		if existing:
			return existing
		record: Reviewer = {
			"id": user_id,
			"addedBy": added_by,
			"addedAt": _now_iso(),
		}
		self.data["reviewers"].append(record)
		self.touch()
		self.flush()
		return record

	async def remove_reviewer(self, user_id: str) -> None:
		self.data["reviewers"] = [r for r in self.data["reviewers"] if r["id"] != user_id]
		self.touch()
		self.flush()

	async def list_reviewers(self) -> list[Reviewer]:
		return list(self.data["reviewers"])

	# Guild state

	async def get_guild_state(self, guild_id: str) -> GuildState | None:
		for state in self.data["guildStates"]:
			if state["id"] == guild_id:
				return state
		return None

	async def set_guild_state(self, guild_id: str, patch: dict[str, object]) -> GuildState:
		for index, state in enumerate(self.data["guildStates"]):
			if state["id"] == guild_id:
				updated: GuildState = {**state, **patch, "id": guild_id}  # type: ignore[typeddict-item]
				self.data["guildStates"][index] = updated
				self.touch()
				self.flush()
				return updated
		record: GuildState = {
			"id": guild_id,
			"antiraidBlocked": False,
			"antiraidBots": [],
			"enforcementStatus": "auto_enforced",
			"antiraidNotifiedAt": None,
			"lastEnforcementAt": None,
			**patch,  # type: ignore[typeddict-item]
		}
		self.data["guildStates"].append(record)
		self.touch()
		self.flush()
		return record

	async def mark_guild_enforced(self, guild_id: str) -> None:
		now = _now_iso()
		for index, state in enumerate(self.data["guildStates"]):
			if state["id"] == guild_id:
				updated: GuildState = {
					**state,
					"enforcementStatus": "manually_enforced",
					"lastEnforcementAt": now,
					"id": guild_id,
				}
				self.data["guildStates"][index] = updated
				self.touch()
				self.flush()
				return
		record: GuildState = {
			"id": guild_id,
			"antiraidBlocked": False,
			"antiraidBots": [],
			"enforcementStatus": "manually_enforced",
			"antiraidNotifiedAt": None,
			"lastEnforcementAt": now,
		}
		self.data["guildStates"].append(record)
		self.touch()
		self.flush()

	async def close(self) -> None:
		if self.loaded:
			self.flush()

	def import_data(self, data: DatabaseSchema) -> None:
		self.data = {**DEFAULT_SCHEMA, **data}
		self.modified = True


def _now_iso() -> str:
	from datetime import datetime

	return datetime.now(UTC).isoformat()

File: src/lib/jsondb.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname } from "node:path";
import type { GuildState, Reviewer, Sanction, SanctionRequest } from "./types";

interface DatabaseSchema {
	sanctionRequests: SanctionRequest[];
	sanctions: Sanction[];
	reviewers: Reviewer[];
	guildStates: GuildState[];
}

const DEFAULT_SCHEMA: DatabaseSchema = {
	sanctionRequests: [],
	sanctions: [],
	reviewers: [],
	guildStates: [],
};

/**
 * Simple JSON-file database for NationSeal. It keeps the entire dataset in
 * memory and flushes to disk after every mutating operation.
 *
 * This is intentionally small-scale: a single-process bot that stores
 * sanctions, reviewer rosters, and guild metadata. For production loads
 * beyond thousands of records, switch back to a real database.
 */
export class JsonDatabase {
	private filePath: string;
	private data: DatabaseSchema = { ...DEFAULT_SCHEMA };
	private flushPromise: Promise<void> | null = null;
	private loaded = false;
	private modified = false;

	constructor(filePath: string) {
		this.filePath = filePath;
	}

	async connect(): Promise<void> {
		if (this.loaded) return;
		try {
			const raw = await readFile(this.filePath, "utf-8");
			const parsed = JSON.parse(raw) as Partial<DatabaseSchema>;
			this.data = { ...DEFAULT_SCHEMA, ...parsed };
			this.modified = false;
		} catch (error) {
			if (typeof error === "object" && error && "code" in error && error.code === "ENOENT") {
				this.data = { ...DEFAULT_SCHEMA };
				this.modified = false;
				await this.flush();
			} else {
				throw error;
			}
		}
		this.loaded = true;
	}

	private async flush(): Promise<void> {
		if (!this.modified) return;
		// Chain writes so two concurrent mutations can't interleave and
		// produce a torn JSON file.
		this.flushPromise = (this.flushPromise ?? Promise.resolve()).then(async () => {
			if (!this.modified) return;
			await mkdir(dirname(this.filePath), { recursive: true });
			await writeFile(this.filePath, JSON.stringify(this.data, null, "\t"), "utf-8");
			this.modified = false;
		});
		await this.flushPromise;
	}

	private touch(): void {
		this.modified = true;
	}

	// Sanction requests

	async createSanctionRequest(data: Omit<SanctionRequest, "id">): Promise<SanctionRequest> {
		const id = crypto.randomUUID();
		const record = { ...data, id } as unknown as SanctionRequest;
		this.data.sanctionRequests.push(record);
		this.touch();
		await this.flush();
		return record;
	}

	async getSanctionRequest(id: string): Promise<SanctionRequest | undefined> {
		return this.data.sanctionRequests.find((r) => r.id === id);
	}

	async updateSanctionRequest(
		id: string,
		patch: Partial<SanctionRequest>,
	): Promise<SanctionRequest> {
		const index = this.data.sanctionRequests.findIndex((r) => r.id === id);
		if (index === -1) {
			throw new Error(`Sanction request ${id} not found`);
		}
		const updated = { ...this.data.sanctionRequests[index], ...patch, id };
		this.data.sanctionRequests[index] = updated as SanctionRequest;
		this.touch();
		await this.flush();
		return updated as SanctionRequest;
	}

	async listSanctionRequestsByStatus(
		status: SanctionRequest["status"],
		limit: number,
	): Promise<SanctionRequest[]> {
		return this.data.sanctionRequests
			.filter((r) => r.status === status)
			.sort((a, b) => new Date(b.submittedAt).getTime() - new Date(a.submittedAt).getTime())
			.slice(0, limit);
	}

	// Sanctions

	async getSanction(targetId: string): Promise<Sanction | undefined> {
		return this.data.sanctions.find((s) => s.id === targetId);
	}

	async upsertSanction(targetId: string, data: Omit<Sanction, "id">): Promise<Sanction> {
		const index = this.data.sanctions.findIndex((s) => s.id === targetId);
		const record = { ...data, id: targetId } as Sanction;
		if (index === -1) {
			this.data.sanctions.push(record);
		} else {
			this.data.sanctions[index] = record;
		}
		this.touch();
		await this.flush();
		return record;
	}

	async removeSanction(targetId: string): Promise<void> {
		this.data.sanctions = this.data.sanctions.filter((s) => s.id !== targetId);
		this.touch();
		await this.flush();
	}

	async listAllSanctions(): Promise<Sanction[]> {
		return [...this.data.sanctions];
	}

	// Reviewers

	async isReviewer(userId: string): Promise<boolean> {
		return this.data.reviewers.some((r) => r.id === userId);
	}

	async addReviewer(userId: string, addedBy: string): Promise<Reviewer> {
		const existing = this.data.reviewers.find((r) => r.id === userId);
		if (existing) return existing;
		const record: Reviewer = {
			id: userId,
			addedBy,
			addedAt: new Date().toISOString(),
		};
		this.data.reviewers.push(record);
		this.touch();
		await this.flush();
		return record;
	}

	async removeReviewer(userId: string): Promise<void> {
		this.data.reviewers = this.data.reviewers.filter((r) => r.id !== userId);
		this.touch();
		await this.flush();
	}

	async listReviewers(): Promise<Reviewer[]> {
		return [...this.data.reviewers];
	}

	// Guild state

	async getGuildState(guildId: string): Promise<GuildState | undefined> {
		return this.data.guildStates.find((g) => g.id === guildId);
	}

	async setGuildState(
		guildId: string,
		patch: Partial<Omit<GuildState, "id">>,
	): Promise<GuildState> {
		const index = this.data.guildStates.findIndex((g) => g.id === guildId);
		let record: GuildState;
		if (index === -1) {
			record = {
				id: guildId,
				antiraidBlocked: false,
				antiraidBots: [],
				enforcementStatus: "auto_enforced",
				antiraidNotifiedAt: null,
				lastEnforcementAt: null,
				...patch,
			} as GuildState;
			this.data.guildStates.push(record);
		} else {
			record = { ...this.data.guildStates[index], ...patch, id: guildId } as GuildState;
			this.data.guildStates[index] = record;
		}
		this.touch();
		await this.flush();
		return record;
	}

	async markGuildEnforced(guildId: string): Promise<void> {
		const index = this.data.guildStates.findIndex((g) => g.id === guildId);
		const patch: Partial<GuildState> = {
			enforcementStatus: "manually_enforced",
			lastEnforcementAt: new Date().toISOString(),
		};
		if (index === -1) {
			this.data.guildStates.push({
				id: guildId,
				antiraidBlocked: false,
				antiraidBots: [],
				enforcementStatus: "manually_enforced",
				antiraidNotifiedAt: null,
				lastEnforcementAt: new Date().toISOString(),
			});
		} else {
			this.data.guildStates[index] = {
				...this.data.guildStates[index],
				...patch,
				id: guildId,
			} as GuildState;
		}
		this.touch();
		await this.flush();
	}

	async close(): Promise<void> {
		if (this.loaded) {
			await this.flush();
		}
	}

	/** Internal helper for one-off migrations: replaces the whole dataset. */
	importData(data: DatabaseSchema): void {
		this.data = { ...DEFAULT_SCHEMA, ...data };
		this.modified = true;
	}
}
